### DISCUSSION TITLE
Preparing for "Web Components Community Group: 2022 Spec/API status"

### DISCUSSION THREAD

**Discussion by Westbrook** [w3c/webcomponents-cg#31] [Ideas]
I'll break out separate threads below for each of the topics available in [our 2021 report ](https://w3c.github.io/webcomponents-cg)so we can gather updates on status and outline what we might want to say about those APIs this year.

If you've got APIs that we chose not to include (even if just for time), please open a new discussion thread below so we can get an idea of what coverage we'd like to make for this years report.

We'll also be gathering for TBD life sessions to discussion these on a weekly or bi-weekly basis leading up to TPAC. Share you availability [here](https://www.when2meet.com/?16122613-fZyNk) if you'd like to join, and checkout our [Slack channel](https://join.slack.com/t/webcomponentcommunity/shared_invite/zt-1cihokx0x-r1_nN8auWIo4~prlQG3SYw) to ensure you're included in invites, etc.

**Comment by Westbrook** [comment]
[Form Associated Custom Elements](https://w3c.github.io/webcomponents-cg/#form-associated-custom-elements)

**Reply by calebdwilliams** [reply]
A lot of really good momentum here. Chrome has had form-associated custom elements turned on for awhile now, but as of around version 100, Firefox has enabled the full `ElementInternals` spec (minus `CustomStateSet`) which comprises the form-associated custom elements features. Safari still doesn't have the feature enabled (but [it can be polyfilled](https://www.npmjs.com/package/element-internals-polyfill)), but there does appear to be [active work on adding `ElementInternals` to WebKit](https://bugs.webkit.org/show_bug.cgi?id=197960).

There are a few small questions where the spec hasn't been implemented entire, specifically [in relation to `formStateRestoreCallback`](https://bugs.chromium.org/p/chromium/issues/detail?id=1335202) which should support autocomplete features, but it's not clear from the spec how exactly that feature should be triggered and based on the source for Chrome and Firefox, neither browser has implemented the feature.

**Reply by claviska** [reply]
A WebKit PR for this is located here: https://github.com/WebKit/WebKit/pull/2690

**Reply by thescientist13** [reply]
For those following along, this was added to our report #42  🙌

**Comment by Westbrook** [comment]
[Cross Root Aria](https://w3c.github.io/webcomponents-cg/#cross-root-aria)

**Reply by Westbrook** [reply]
Some great progress has come together on Attribute Delegation via: https://github.com/leobalter/cross-root-aria-delegation and its corollary Attribute Reflection (name is a work in progress) via: https://github.com/Westbrook/cross-root-aria-reflection. The earlier has pretty solid agreement from browsers and is looking like it's a spec text approval away from beginning to gather implementors. The later is newer, but based off of similar ideas and holding similar excitement, if not yet agreement, from vendors.

Together, they beg a question as to whether this should be an "Aria" spec or an "Attribute" spec, so that we can easily pass references for concepts including those outlined by the [Popup API](https://open-ui.org/components/popup.research.explainer) and [CSS Anchor Positioning](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/CSSAnchoredPositioning/explainer.md).

**Reply by Westbrook** [reply]
If you have thoughts on the [growing explainer](https://github.com/Westbrook/cross-root-aria-reflection/blob/main/cross-root-aria-reflection.md) for Aria Attribute Reflection, would love to see issues or questions here so we can continue to flesh it and the related Aria Attribute Delegation proposals out for implementors.

**Reply by Westbrook** [reply]
Include Element references in this convo.

**Reply by thescientist13** [reply]
For those following along, PR us up #46 🙌

**Comment by Westbrook** [comment]
[Constructible Style Sheets & adoptedStyleSheets](https://w3c.github.io/webcomponents-cg/#constructable-stylesheets-adoptedstylesheets)

**Reply by Westbrook** [reply]
Firefox shipped both of these in v101 this year. This leaves only Safari missing an implementation.

**Reply by Westbrook** [reply]
No longer frozen array.

**Reply by thescientist13** [reply]
Added an issue to the [Webkit Standards Position repo](https://github.com/WebKit/standards-positions/issues/42)

edit:

got some feedback comments there already that I'm not super familiar with, but will brush up on those referenced issues and see if I can comment back.

**Reply by thescientist13** [reply]
PR is up for those following along #50 🙌

**Comment by Westbrook** [comment]
[CSS Module Scripts](https://w3c.github.io/webcomponents-cg/#css-module-scripts)

**Reply by Westbrook** [reply]
Should we push on this as "Import Assertions", _generally_, or is there a benefit to focus specifically on CSS at this point?

**Reply by calebdwilliams** [reply]
This depends at least in part on constructible stylesheets, so that should take precedence. I think that pushing on import assertions generally would do us well here especially because CSS modules are likely a fast-follow to that and will depend on it.

This will likely also have an effect on a lot of the tools developers use day-to-day (TypeScript supports the syntax, Rollup does too with the right Acorn plugins, Webpack isn't super happy about it, I don't know about others), so broadening the scope of the effort here will be worthwhile.

**Reply by Westbrook** [reply]
That's an interesting angle to take (in this report or another), the community's support for various APIs/spec via tooling.

When a polyfill is appropriate, it would probably be good to list it in the report, aligns closely to Web Platform Test that we might take some role in making sure exist and are referenced appropriately. When you actually have to process code for it to work, see `assert` then an apendix or related report might be my suggestion. Thoughts?

**Reply by calebdwilliams** [reply]
I think that makes sense.

Also, many of the technologies in this document are related at some level of scale. It might also do us well to map those relationships out explicitly.

**Comment by Westbrook** [comment]
[Scoped Element Registries](https://w3c.github.io/webcomponents-cg/#scoped-element-registries)

**Reply by thescientist13** [reply]
Opened an issue for tracking over in the new Safari Tracker - https://github.com/WebKit/standards-positions/issues/38

**Reply by Westbrook** [reply]
@justinfagnani or @leobalter did push in this area from the team at Sales Force get us any closer to proposed spec text?

**Reply by justinfagnani** [reply]
Nothing yet, but hopefully

**Reply by Westbrook** [reply]
Is this the NUMBER 1 priority now that other things have started to ship a little?

How can the DSD interaction that Safari is worried about be clarified?

**Reply by Westbrook** [reply]
Let's open PRs for all the issues so we can move this forward!

**Reply by thescientist13** [reply]
PR open / merged for anyone following along - #51  🙌

**Comment by Westbrook** [comment]
[Declarative Shadow DOM](https://w3c.github.io/webcomponents-cg/#declarative-shadow-dom)

**Reply by thescientist13** [reply]
From my understanding I don't think much has changed from the standards side in terms of movement; right now only Chromium based browsers (Chrome, Edge) support Declarative Shadow DOM?

For Safari, it appears that landing DSD is very [dependent on Scoped Custom Element Registries](), which happens to be one our [top items](https://github.com/w3c/webcomponents-cg/discussions/31#discussioncomment-3130117) as well.  I don't see a proposal for it in the Safari tracker so will do so now to get it on their radar, though hopefully we can find a way to tackle them independently.

From a community perspective I'm not sure if any [real world use cases](https://github.com/whatwg/dom/issues/831#issuecomment-993902543) have popped up that could help sway hearts and minds?  So if we can find those examples and / or maybe highlight community projects leveraging Declarative Shadow DOM in some way to help showcase enthusiasm / interest?

---

edit: 

Opened an issue for tracking Scoped Custom Element Registries on the Safari tracker - https://github.com/WebKit/standards-positions/issues/38

**Reply by calebdwilliams** [reply]
For a real-world use case, I believe Reddit is doing a rewrite using Lit and SSR. Might be worth reaching out there.

**Reply by koddsson** [reply]
> For a real-world use case, I believe Reddit is doing a rewrite using Lit and SSR. Might be worth reaching out there.

Is this written about somewhere? I'd love to read up on their use case and how they came to this conclusion.

**Reply by otabekshoyimov** [reply]
> For a real-world use case, I believe Reddit is doing a rewrite using Lit and SSR. Might be worth reaching out there.

Can you a link to blog, video, twitter, reddit thread about this info?

**Reply by otabekshoyimov** [reply]
*give

**Reply by sashafirsov** [reply]
@thescientist13 
> not sure if any real world use cases have popped up

I have been looking into good show case for [Declarative Shadow DOM](https://w3c.github.io/webcomponents-cg/#declarative-shadow-dom) as part of its evolution in [shadow-dom-element](https://github.com/sashafirsov/shadow-dom-element). The static site template  [link](https://unpkg.com/shadow-dom-element@0.0.7/demo/ananke/content/en/index.html) came as first. But to be attractive it has to come with well developed theme which is quite time consuming.  

The second choice is the template for json-resume. My bet was in interest of developer to publish their CV in well designed and responsive format with flexible customization via per-section overridable templates. Work on `json-resume` component is in progress.

**Reply by thescientist13** [reply]
Thanks for the info @sashafirsov .  But just to clarify, the definition of "real world use cases" I was citing was this
> _is to get real-world **production** site performance data, showing an optimized site in the wild. It is [relatively easy](https://github.com/mfreed7/declarative-shadow-dom/blob/master/perf_tests/clock_example/results.md) to gin up example sites that either make things look good or bad, but the question is whether there's a real world performance benefit (or penalty) for using DSD, **when used in production**._

So not to say your work isn't valuable to share as in the next line they do acknowledge and appreciate this kind of work, but just from the context of this next draft of our report, hopefully it clarifies what examples we'd be looking to include / cite.

----

That said, I do acknowledge its still a bit of chicken / egg situation, asking for real world use cases in a chromimum only world, but maybe making these tools to make it easier  for someone else to go to production can be the inspiration for someone to give it a go anyway!  Let's see if we can get some folks to build some sites!! 💯 🤞

**Reply by sashafirsov** [reply]
As `template mode` is only Chrome feature, I doubt it would be adopted without wider support and proper polyfill. Also this feature without external templates and data and slots passed as attributes is not sufficient for real world apps.  Its born dead due to lack of MVP.

**Reply by thescientist13** [reply]
There is [a polyfill](https://web.dev/declarative-shadow-dom/#polyfill), FYI.  I have used it and can confirm it works.

Anyway, as I mentioned I do acknowledge the chicken / egg situation and understand it makes taking a chance on the feature hard by a "stakeholder" but it is where we are at, so just trying to make the best of it by hoping to find a way for implementers and consumers to meet in the middle.  ➡️ ⬅️

**Reply by koddsson** [reply]
We at GitHub are interested in the declarative shadow dom, but since there's no broad browser support, we have been hesitant to ship it to production. The polyfill is excellent, but since it falls back to normal shadow DOM, we won't render anything if the JS for the component doesn't execute.

**Reply by Westbrook** [reply]
@koddsson how does GH handle no JS today? Should the community be looking into hybrid solutions where there’s a DSD fallback from JS and a no-DSD fallback from there? In most of the apps I work on you _have_ to have JS at some point and it would be a loading process benefit. However, in a form processable application like GH a I could see the benefits there.

**Reply by Westbrook** [reply]
Not the most beautiful to write solution, but possibly an interesting path with the support of tooling, if you're really looking to support no-JS contexts: https://codepen.io/Westbrook/pen/bGvaOQP

**Reply by mfreed7** [reply]
@koddsson I'm excited to hear GitHub is interested in DSD! That would likely help convince the other implementers of the value of the feature overall, especially if any public performance data comes out of that effort. Let me know if I can help in any way.

Side note: I've worked with a few internal partners who were planning to just **not** ship a polyfill with their SSR content. In that case, the `<template shadowroot=open>` element falls back to an inert `<template>` on unsupported rendering engines, and the page waits for JS to hydrate the shadow roots when they load, as they do today. Perhaps that's an option?

**Reply by koddsson** [reply]
> @koddsson how does GH handle no JS today? Should the community be looking into hybrid solutions where there's a DSD fallback from JS and a no-DSD fallback from there? In most of the apps I work on you _have_ to have JS at some point and it would be a loading process benefit. However, in a form processable application like GH a.

The unofficial stance is to try to make as much as possible of the site work in "read-only" mode when JS doesn't load. That's more of guidance, so some teams write entirely JS-rendered experiences. I think we can start experimenting with DSD rendering in those places.

> @koddsson I'm excited to hear GitHub is interested in DSD! That would likely help convince the other implementers of the value of the feature overall, especially if any public performance data comes out of that effort. Let me know if I can help in any way.

I'll brainstorm on this a bit. We've been cautious about implementing just the Shadow DOM to make sure we aren't breaking things for some users, but recently we shipped Shadow DOM in [`<image-crop>`](https://github.com/github/image-crop-element) and [`<include-fragment>`](https://github.com/github/include-fragment-element). Maybe we should start looking at places where we can test out DSD. cc/ @keithamus

What should we be looking out for in terms of performance? Maybe I can start monitoring this week, so we have some data to compare to when we ship DSD.

**Reply by keithamus** [reply]
The hesitancy for us right now is lack of strong positive signals from Mozilla or Apple.

The performance of ShadowDOM in my opinion is an overall less interesting story than the architectural benefits. One of the big issues we face with our design system today is encapsulated CSS, which ShadowDOM effortlessly fixes, along with other ergonomic improvements such as less complex class names, and flexibility around CSS variables penetrating shadow roots, light dom styles etc.

I'm confident that if we migrated our components _entirely_ to using ShadowDOM we could improve LCP overall as one of the biggest costs to our LCP is the CSS bundles we have. DSD takes that up a notch as we're able to move reification of shadowroots before DOMContentLoaded, but largely it feels like we'd gain the most performance from componentising our CSS using ShadowDOM. The concern around just using ShadowDOM is the lack of resiliency: without JS the site fails to do anything useful, and that to us is a threshold we're not prepared to cross in the general case. As @koddsson points out we have a concept of "read only" mode, where we expect a baseline experience to exist without JS. Therefore without a clear path to having DSD implemented in all browsers, where we can move wholesale to reap the rewards, we're stuck in a state where some (Firefox & Safari) users will get a slower and less resilient experience, for the benefit of (currently) Chrome & Edge users.

We also need to think about the long tail of users on older browsers, where DSD much less ShadowDOM doesn't exist, however can be solved by simply waiting. Performance issues aside, we're happy enough with the current state of DSD polyfills, but ShadowDOM polyfills (like ShadyDOM) cause severe performance degradation on the site. Up until very recently we've still had a core contingent of browsers that do not support ShadowDOM which has been a blocker, but that's being lifted.

With regards to Scoped Element Registries - this isn't a concern for us. Our highest priority would be DSD across all browsers.

**Reply by mfreed7** [reply]
@keithamus thanks for the detailed explanation about GitHub's motivation for shadow DOM and DSD! This would be a **very** helpful comment to add on https://github.com/whatwg/dom/issues/831, as it goes to the need for the feature, independent of performance concerns.

**Reply by thescientist13** [reply]
Opened a PR to get this added to our report - https://github.com/w3c/webcomponents-cg/pull/45.

**Comment by Westbrook** [comment]
[Composed Selection](https://w3c.github.io/webcomponents-cg/#composed-selection)

**Reply by hunterloftis** [reply]
- @mfreed7 has been driving progress since last year
- TAG [concluded design review & approves](https://github.com/WICG/webcomponents/issues/79#issuecomment-1159112560)
- I'm not aware of any remaining objections / use-cases that are not possible with the [current proposed design](https://github.com/mfreed7/shadow-dom-selection/blob/main/README.md) of `getComposedRange()`
    - Although only one of the major rich text editors [from our report](https://w3c.github.io/webcomponents-cg/#composed-selection) has responded:
        - [TinyMCE](https://github.com/tinymce/tinymce/pull/4737#issuecomment-966097404)
        - CKEditor: crickets
        - Quill: crickets
- I'm not aware of any movement of browser implementations (other places to check? If not, where to nudge?)
    - [5-year old Mozilla issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1430308)
    - [6-year old Webkit issue](https://bugs.webkit.org/show_bug.cgi?id=163921)
- I *think* that the current **Status** should be updated to:
    - General consensus, no implementation of the agreed-upon design, divergent implementations of the previous selection API

**Comment by Westbrook** [comment]
[Declarative CSS Modules](https://w3c.github.io/webcomponents-cg/#declarative-css-modules)

**Reply by calebdwilliams** [reply]
Calling out for the sake of completeness, this depends in part on constructible stylesheets (or `DocumentOrShadowRoot.prototype.adoptedStyleSheets`), CSS module scripts and likely import assertions.

**Reply by hsablonniere** [reply]
This one is very important :pray:
Without this, Declarative Shadow DOM feels a bit half baked as for each instance of a component, we need to repeat the same CSS in the rendered output.
Combined with CSS Module Scripts, this would be a nice standard way to render web components from a server.

**Reply by nolanlawson** [reply]
You can use a `<link rel=stylesheet>` to avoid some repetition (vs `<style>`), but yeah, it's not ideal.

For us (LWC), it's a bit awkward to have constructable stylesheets rendered on the client but `<style>`s/`<link>`s in SSR.

**Reply by sashafirsov** [reply]
Seems like the need to share styles across subset of components is a must requirement for WS library. Tried to emphasize the importance of `namespace` scope in #scoping-wc channel. Such namespacing is needed not just for CSS but also for scoped registry, module loader, etc.

**Comment by Westbrook** [comment]
[Children changed callback](https://w3c.github.io/webcomponents-cg/#children-changed-callback)

**Reply by Westbrook** [reply]
Might interact with Imperative Slotting?

Might interact with https://github.com/WICG/webcomponents/issues/933

**Reply by alangdm** [reply]
I tried gathering the info for this, there wasn't much consensus on anything but hopefully it helps to show the current situation:

<details>
<summary>Children Changed Calback</summary>

# Children Changed Callback

## Description 

Provide a callback that triggers whenever the parser finishes parsing children or when a new child is inserted or an old child is removed.
 
## Links

- Discussions:
  - [2019 discussion - open](https://github.com/WICG/webcomponents/issues/809)
  - [December 2016 discussion - closed](https://github.com/WICG/webcomponents/issues/619)
  - [August 2016 discussion regarding parsing timing - closed](https://github.com/WICG/webcomponents/issues/551)
  - [August 2016 discussion - closed](https://github.com/WICG/webcomponents/issues/550)
- Polyfills/Libraries:
  - [For the parsing behavior](https://github.com/WebReflection/html-parsed-element)

## Status 
- No concrete proposal / no consensus on spec

## Initial API Summary/Quick API Proposal

```js
class ChildrenChangedCallbackSample extends HTMLElement {
  // couldn't find any reference to any proposed function signature
  childrenChangedCallback() {}
}
```

## Key Scenarios 

- **Children parsed**: This proposal would act as a cross-browser method of detecting when a custom element's children have been fully parsed reliably.
- **Children inserted/removed**: This proposal would provide better ergonomics to react to insertion or removal of children nodes than using a MutationObserver.

## Concerns

1. Too expensive to implement - no longer the case, all engines have mechanisms for this already
2. The addition of the parsing behavior might encourage assuming a set of initial children exists and not reacting to changes in them.

## Dissenting Opinion 

1. Some people believe using a MutationObserver is enough to cover the insertion/removal cases and want to avoid creating a duplicate similar API for custom elements. But the async nature of MutationObserver seems to cause timing issues.
2. Some other people don't want to add this API if it will only cover the use case when the parser finishes parsing children. (See concern 2)

## Related Specs

- [`slotchange` event options](https://github.com/WICG/webcomponents/issues/933)

## Open Questions

- Should this API really cover both use cases to begin with? (Arguments for both sides have been mentioned in all referenced discussions with no conclusion)
- Should this be one or two separate callback functions?
- Would it be better to implement this behavior as [part of `slotchange` as suggested in this other issue](https://github.com/WICG/webcomponents/issues/933)? - this might lead to engine security bugs so the callback might be preferred
</details>

**Reply by alangdm** [reply]
I created a PR for this so we can do the review process there https://github.com/w3c/webcomponents-cg/pull/40

**Reply by thescientist13** [reply]
For those following along, this was added to our report #40  🙌

**Comment by Westbrook** [comment]
[Open styling of shadow roots](https://w3c.github.io/webcomponents-cg/#open-styling-of-shadow-roots)

**Reply by Westbrook** [reply]
@davatron5000 like these...can we get him to own it?

**Reply by davatron5000** [reply]
Sorry, I can't. I can barely attend meetings.

**Reply by Westbrook** [reply]
Understand and preparation for this situation may benefit from https://github.com/Devographics/Surveys/issues/19?

**Reply by Westbrook** [reply]
Does this conversation benefit from pairing with the `@scope` specification work?

**Reply by Westbrook** [reply]
> heard this being proposed when developers want to continue to use global styles (like typography styles) across document and multiple shadow roots.

**Reply by davatron5000** [reply]
FWIW, I was just asked today "How do you handle shared styles?" by someone getting into web components. The styling story here isn't great.

**Reply by thescientist13** [reply]
Might be worth getting feedback and anecdotes captured in this docs + guides repo - https://github.com/webcomponents-cg/docs-and-guides/issues/21?  Or capture somewhere for sure but I agree there is a lot of opportunity in knowledge sharing / documenting around CSS and WCs, especially since CSS is global by default and so that (naturally) dominates the implementation assumptions of the CSS ecosystem at large.

IMO, it would really be great to have a dedicated guide on styling WCs, either globally and / or with (Declarative) Shadow DOM.

**Comment by Westbrook** [comment]
[Declarative custom elements](https://w3c.github.io/webcomponents-cg/#declarative-custom-elements)

**Reply by sashafirsov** [reply]
The declarative custom element would need dependencies management protocol. Which most likely would match the HTML module model. 
@thescientist13 , if this protocol a part of HTML module concept, should we describe it in a separate proposal?

**Reply by sashafirsov** [reply]
Created #32 to discuss and track it further

**Reply by thescientist13** [reply]
For those following along, this was added to our report #41  🙌

**Comment by Westbrook** [comment]
Themeing (::theme)

GitHub Issues: [WICG/webcomponents#864](https://github.com/WICG/webcomponents/issues/864)

**Reply by alangdm** [reply]
I'm working on this

**Reply by thescientist13** [reply]
Woo hoo, PR is up and ready for review - #54 🙌

**Comment by Westbrook** [comment]
Styling children of slotted content

refs #5

**Comment by Westbrook** [comment]
Custom CSS State

GitHub Issues: https://github.com/WICG/webcomponents/issues/738

**Reply by Westbrook** [reply]
Container Style Queries seem to be giving the capabilities here a run for their money... https://una.im/style-queries/#4-grouping-styles-with-higher-order-variables

**Reply by thescientist13** [reply]
PR for this is up - #56 ! 🙌

**Comment by Westbrook** [comment]
Imperative Slot Assignment

re: https://github.com/whatwg/html/issues/3534

**Reply by Westbrook** [reply]
Only Safari seems to be missing an implementation here.

**Reply by thescientist13** [reply]
@Westbrook 
Not sure if I am reading this right but for Safari, I do see an item for _Imperative Slot API_ that links to this DOM spec in their feature tracker, which seems to be the same thing being discussed in https://github.com/whatwg/html/issues/3534? 
<img width="1331" alt="Screen Shot 2022-09-04 at 2 28 36 PM" src="https://user-images.githubusercontent.com/895923/188328838-4f4939d2-db64-418b-a66b-57a359182341.png">

Or is there a nuance I'm missing around this spec?  Here is what I have so far
- Chrome - https://chromestatus.com/feature/5711021289242624
- Firefox - https://github.com/mozilla/standards-positions/issues/409
- Safari - https://webkit.org/status/#feature-imperative-slot-api

**Reply by Westbrook** [reply]
I’m pretty sure they recently “finished it”, but it’s unclear whether they’ve “shipped it”, past I checked.

**Reply by Westbrook** [reply]
For instance this doesn’t work in iOS, but might work in desktop/tech preview: https://codepen.io/Westbrook/pen/GRvbWzM

**Reply by thescientist13** [reply]
Ah I see.  gotcha.  That helps, as I wasn't sure how to frame the details in regards to the report, but should be able to make a go of it now.

**Reply by thescientist13** [reply]
For those following along, PR is up - https://github.com/w3c/webcomponents-cg/pull/55!  🙌

**Comment by Westbrook** [comment]
CSS Properties and values inside shadow root

re: https://bugs.chromium.org/p/chromium/issues/detail?id=1231135

**Comment by Westbrook** [comment]
Lazy custom element definitions

GitHub Issues: https://github.com/WICG/webcomponents/issues/782

**Reply by thescientist13** [reply]
For those following along, this was added to our report #41  🙌

**Comment by Westbrook** [comment]
DOM Parts (formerly Template Instantiation)

GitHub Issues: https://github.com/WICG/webcomponents/blob/gh-pages/proposals/DOM-Parts.md

**Reply by sashafirsov** [reply]
Added as todo item to implement in `css-chain` and derived `shadow-dom-element` and `light-dom-element` ( kind of declarative custom element polyfills ). Participation on browser Chromium/Blink and FF level is problematic due to time resources limits.

**Reply by sashafirsov** [reply]
Need someone to hand off or pair on this proposal status.

**Reply by thescientist13** [reply]
For those following along, this was added to our report #41  🙌

**Comment by Westbrook** [comment]
HTML modules

GitHub Issues: https://github.com/WICG/webcomponents/issues/645

**Reply by Westbrook** [reply]
Without Constructible Style Sheets you can't have CSS Module Scripts, you import something that you assert to be CSS and you get a Constructible Style Sheet. This _might_ not be "HTML Module", but what is we could assert that the thing we were importing was a custom element or "web component" and the thing we got was an element definition in its own module scope? In that way, all of the the custom elements defined within that scope could be encapsulated from the rest of the document, giving us an alternative path to Scoped Custom Elements.

It does sort of imply that if you delivered DSD into a custom element that you want to import as a Custom Element Module later we'd need to work out upgrading the shadow DOM to that scope. This seems like something you'd need either way. Either upgrading to include a new scope or adding to the DSD spec to register a scope and then be able to acquire the scope from the shadow root at a later time. This may be specifically the question at hand for Safari in regard to DSD and Scoped Registries, so it'll be interesting to see how that conversation goes.

**Reply by thescientist13** [reply]
PR for HTML Modules in the 2022 report is up - #60!

**Comment by Westbrook** [comment]
Native Behavior Mixins

link to conversation?

**Reply by Westbrook** [reply]
Some conversations around alternatives to customized built-ins point to the need for behavior mixins. Thinks like the `for` functionality in a `<label>` or `showModal()` functionality in a `<dialog>` or the `behavior` attribute in the `<selectmenu>` spec…

**Reply by LeaVerou** [reply]
This is actually useful for regular web components as well. Often a component is twopart: One or more new elements plus attributes to augment existing elements. Think of `<datalist>` and the `<input list>` attribute.
Relevant: https://github.com/whatwg/html/issues/2271
It would be good to have an API for easily registering custom attributes (and maybe synced properties?) for other elements with code to react to changes (yes, we can do that with a Mutation Observer that listens to the entire document subtree, but maybe there's a better way).

**Reply by calebdwilliams** [reply]
`MutationObserver` is less than ideal because you'd also have to watch all shadow trees which is less than ideal. Curious if custom attributes would also need to be scoped like custom element registries.

**Reply by calebdwilliams** [reply]
It would be worthwhile to identify a list of built-in features we'd like to be able to add to the custom elements spec. Creating labelable elements is a great example. Creating a form-associated element (with or without attaching an internals instance) allows an element to accept a label, so 

```html
<script>
customElements.define('x-', class X extends HTMLElement {
  static formAssociated = true;
});
</script>
<form>
  <label for="x">This works via the AOM</label>
  <x- id="x"></x->
</form>
```

will create a labelable element, but that element does show up in `HTMLFormElement.prototype.elements` collection which is not always what a user will want when creating a loader, for one example.

It could be that just adding additional static props would be sufficient for many of these approaches.

**Reply by Westbrook** [reply]
The attribute relationship here plays a role in the [Aria delegate/reflect APIs](https://github.com/Westbrook/cross-root-aria-reflection/blob/main/cross-root-aria-reflection.md#:~:text=%23shadow%2Droot%20delegates%3D%22for%22) that may appease some of the use case there. However, the delegation approach there assumes the presence of a `<label>` element to complete the contract. I'm interested in what it would mean to make an `<x-label>` element that never relied on a native `<label>` element to fulfill the responsibilities that it needs to fill. So, instead of `class XLabel extends HTMLLabelElement {}` you could do something like `class XLabel extends WithForAttribute(HTMLElement) {}` or something like:

```
class XLabel extends HTMLElement {
  constructor() {
    super();
    this.#internals = this.attachInternals();
    this.#internals.mixins = [ withForElement, ... ];
}
```

Or something else along those lines.

**Comment by mfreed7** [comment]
> I'll break out separate threads below for each of the topics available in [our 2021 report ](https://w3c.github.io/webcomponents-cg)so we can gather updates on status and outline what we might want to say about those APIs this year.

@Westbrook I've got a meta-question about the 2022 report. The ["2021 version"](https://w3c.github.io/webcomponents-cg/) is located at a base URL (https://w3c.github.io/webcomponents-cg/) that doesn't have "2021" in the name, and it has been updated as of 4 days ago (18 August 2022). Is there a plan to keep static annual reports ("2021", "2022", etc.) or is the plan to have a living document that is continuously updated? I'd vote for the former (distinct annual reports), since it gives us (implementers) a stake in the ground and allows us to reference changes over time as we implement things. The living document approach can also be useful, if you'd prefer to go that route. But in that case, could you please change the name accordingly? E.g. "Web Components Community Group: Evergreen Spec/API status" or something better.

Side note, which I've said multiple times: this effort to organize and prioritize outstanding Web Components issues is **amazingly useful**! **Thank you all for putting in the effort!**

**Reply by Westbrook** [reply]
@mfreed7 this is a great question and while we do have plans for yearly reports there definitely seems to be some planning missing in said plan. I've just been using the rSpec tooling from w3c without much thought as to when it says things have "updated" or not, see what I get for just assuming 🙈. The 2021 report hasn't _technically_ updated since our presentation last year, except to get extra Mozilla position links and get moved to 2021.html in our repo 14 days ago: https://github.com/w3c/webcomponents-cg/commits/main/reports/status.html I'm not sure how to keep the project from continuing to say that it's updated on running publication of the [2022](https://w3c.github.io/webcomponents-cg/2022.html) work in progress report, but I'll see what I can do there.

I'm glad to hear that the "a stake in the ground" approach is useful to you as an implementor, as we as consuming developers are interested in making a more declarative stance around that "stake", but it's still in development. The idea of a Web Components 2022 to stand next to an es2022 feels pretty nice, and kicking of a conversation with implementors around how we could make that productive for everyone would be very nice!

**Reply by mfreed7** [reply]
> @mfreed7 this is a great question and while we do have plans for yearly reports there definitely seems to be some planning missing in said plan. I've just been using the rSpec tooling from w3c without much thought as to when it says things have "updated" or not, see what I get for just assuming 🙈. The 2021 report hasn't _technically_ updated since our presentation last year, except to get extra Mozilla position links and get moved to 2021.html in our repo 14 days ago: https://github.com/w3c/webcomponents-cg/commits/main/reports/status.html I'm not sure how to keep the project from continuing to say that it's updated on running publication of the [2022](https://w3c.github.io/webcomponents-cg/2022.html) work in progress report, but I'll see what I can do there.

Thanks, but I don't think that needs heroics. As long as the title and the content are actually per-year, I think that's great. 


> 
> I'm glad to hear that the "a stake in the ground" approach is useful to you as an implementor, as we as consuming developers are interested in making a more declarative stance around that "stake", but it's still in development. The idea of a Web Components 2022 to stand next to an es2022 feels pretty nice, and kicking of a conversation with implementors around how we could make that productive for everyone would be very nice!

Cool, thanks! This is indeed useful to us.

**Comment by thescientist13** [comment]
For all those following along, our report will be presented this **_Wednesday, Sept. 14th @ 4:30 PST_**
https://www.w3.org/events/meetings/2f1bb21d-1d51-4528-9729-9da2bc611176

**Comment by Westbrook** [comment]
Time to get ready for [2023](https://github.com/w3c/webcomponents-cg/discussions/66), closing. Thanks for all the discussion, I hope you'll all join us in preparation for this year's report.
